homeassistant/components/dlna_dms/dms.py

```diff
@@ -0,0 +1,715 @@
+
+        for resource in item.res:
+            if not resource.uri:
+                continue
+            if mime_type := _resource_mime_type(resource):
+                url = self._device.get_absolute_url(resource.uri)
+                LOGGER.debug("Resolved to url %s MIME %s", url, mime_type)
+                return DidlPlayMedia(url, mime_type, item)
+
+        LOGGER.debug("Object %s has no playable resources", item.id)
+        raise Unresolvable("Object has no playable resources")
+
+    def _didl_to_media_source(
+        self,
+        item: didl_lite.DidlObject,
+        browsed_children: DmsDevice.BrowseResult | None = None,
+    ) -> BrowseMediaSource:
+        """Convert a DIDL-Lite object to a browse media source."""
+        children: list[BrowseMediaSource] | None = None
+
+        if browsed_children:
+            children = [
+                self._didl_to_media_source(child)
+                for child in browsed_children.result
+                if isinstance(child, didl_lite.DidlObject)
+            ]
+
+        # Can expand if it has children (even if we don't have them yet), or its
+        # a container type. Otherwise the front-end will try to play it (even if
+        # can_play is False).
+        try:
+            child_count = int(item.child_count)
+        except (AttributeError, TypeError, ValueError):
+            child_count = 0
+        can_expand = (
+            bool(children) or child_count > 0 or isinstance(item, didl_lite.Container)
+        )
+
+        # Can play if item has any resource that can be streamed over the network
+        can_play = any(_resource_is_streaming(res) for res in item.res)
+
+        # Use server name for root object, not "root"
+        title = self.name if item.id == ROOT_OBJECT_ID else item.title
+
+        mime_type = _resource_mime_type(item.res[0]) if item.res else None
+        media_content_type = mime_type or MEDIA_TYPE_MAP.get(item.upnp_class, "")
+
+        media_source = BrowseMediaSource(
+            domain=DOMAIN,
+            identifier=self._make_identifier(Action.OBJECT, item.id),
+            media_class=MEDIA_CLASS_MAP.get(item.upnp_class, ""),
+            media_content_type=media_content_type,
+            title=title,
+            can_play=can_play,
+            can_expand=can_expand,
+            children=children,
+            thumbnail=self._didl_thumbnail_url(item),
+        )
+
+        media_source.calculate_children_class()
+
+        return media_source
+
+    def _didl_thumbnail_url(self, item: didl_lite.DidlObject) -> str | None:
+        """Return absolute URL of a thumbnail for a DIDL-Lite object.
+
+        Some objects have the thumbnail in albumArtURI, others in an image
+        resource.
+        """
+        assert self._device
+
+        # Based on DmrDevice.media_image_url from async_upnp_client.
+        if album_art_uri := getattr(item, "album_art_uri", None):
+            return self._device.get_absolute_url(album_art_uri)
```

Comment:
Not all absolute URLs can be resolved. For example, if you're outside the network. 

Plex has solved this by adding it's own web view to fall back to when the request is not from the internal network: https://github.com/home-assistant/core/pull/66702/files#diff-4bc5cfdbcf43f8ad35f24f9c6ae347141b2bd4ee85a26bc372b411422e52337bR79-R86

This could be left for an extra PR.

Reply:
Is it worth doing the browser thumbnails proxying in `media_player`, the same way it proxies the currently playing media's thumbnail?

Reply:
Media player doesn't know if a URL can be locally resolved or not.

Reply:
Ah, that makes sense. I'll work on this in a future PR.